flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleServiceFactory.java

```diff
@@ -119,13 +121,33 @@ static NettyShuffleEnvironment createNettyShuffleEnvironment(
                         config.networkBufferSize(),
                         config.getRequestSegmentsTimeout());
 
+        // we create a separated buffer pool here for batch shuffle instead of reusing the network
+        // buffer pool directly to avoid potential side effects of memory contention, for example,
+        // dead lock or "insufficient network buffer" error
+        BatchShuffleReadBufferPool batchShuffleReadBufferPool =
+                new BatchShuffleReadBufferPool(
+                        config.batchShuffleReadMemoryBytes(), config.networkBufferSize());
+
+        // we create a separated IO executor pool here for batch shuffle instead of reusing the
+        // TaskManager IO executor pool directly to avoid the potential side effects of execution
+        // contention, for example, too long IO or waiting time leading to starvation or timeout
+        BatchShuffleReadIOExecutor batchShuffleReadIOExecutor =
+                new BatchShuffleReadIOExecutor(
+                        Math.max(
```

Comment:
maybe let's do this min and max in the ctor of BatchShuffleReadIOExecutor? Then `BatchShuffleReadIOExecutor.MIN_NUM_THREADS` and `BatchShuffleReadIOExecutor.MAX_NUM_THREADS` can be private.